k2/csrc/fsa_utils.cu

```diff
@@ -0,0 +1,327 @@
+    Arc arc{};
+    if (num_fields == 5u) {
+      //   0           1         2         3        4
+      // src_state  dest_state  symbol  aux_label score
+      ++num_arcs;
+      arc.src_state = StringToInt(splits[0]);
+      arc.dest_state = StringToInt(splits[1]);
+      arc.symbol = StringToInt(splits[2]);
+      arc.score = scale * StringToFloat(splits[4]);
+    } else if (num_fields == 1u) {
+      final_state = StringToInt(splits[0]);
+      if (static_cast<int32_t>(state_arcs.size()) <= final_state)
+        state_arcs.resize(final_state + 1);
+      state_aux_labels.resize(final_state + 1);
+      break;  // finish reading
+    } else {
+      K2_LOG(FATAL) << "Invalid line: " << line
+                    << "\nIt expects a line with 5 fields";
+    }
+    if (static_cast<int32_t>(state_arcs.size()) <= arc.src_state) {
+      state_arcs.resize(arc.src_state + 1);
+      state_aux_labels.resize(arc.src_state + 1);
+    }
+
+    state_arcs[arc.src_state].emplace_back(arc);
+    state_aux_labels[arc.src_state].push_back(StringToInt(splits[3]));
+  } while (std::getline(is, line));
+
+  K2_CHECK(is) << "Failed to read";
+  K2_CHECK_NE(final_state, -1) << "Found no final_state!";
+
+  std::vector<int32_t> labels;
+  labels.reserve(num_arcs);
+  for (const auto &v : state_aux_labels) {
+    labels.insert(labels.end(), v.begin(), v.end());
+  }
+
+  *aux_labels = Array1<int32_t>(GetCpuContext(), labels);
+
+  CheckStateArcs(state_arcs, final_state);
+  return AcceptorFromStateArcs(state_arcs, num_arcs);
+}
+
+Fsa FsaFromString(const std::string &s, bool negate_scores /*= false*/,
+                  Array1<int32_t> *aux_labels /*= nullptr*/) {
+  std::istringstream is(s);
+  std::string line;
+  std::getline(is, line);
+  K2_CHECK(is);
+
+  std::vector<std::string> splits;
+  SplitStringToVector(line, kDelim, &splits);
+  auto num_fields = splits.size();
+  if (num_fields == 4u)
+    return AcceptorFromStream(std::move(line), is, negate_scores);
+  else if (num_fields == 5u)
+    return TransducerFromStream(std::move(line), is, negate_scores, aux_labels);
+
+  K2_LOG(FATAL) << "Expected number of fields: 4 or 5."
+                << "Actual: " << num_fields << "\n"
+                << "First line is: " << line;
+
+  return Fsa();  // unreachable code
+}
+
+std::string FsaToString(const Fsa &fsa, bool negate_scores /*= false*/,
+                        const Array1<int32_t> *aux_labels /*= nullptr*/) {
+  K2_CHECK_EQ(fsa.NumAxes(), 2);
+  K2_CHECK_EQ(fsa.Context()->GetDeviceType(), kCpu);
+  const Array1<int32_t> &row_splits = fsa.shape.RowSplits(1);
+  const Array1<Arc> &arcs = fsa.values;
+
+  const int32_t *p = nullptr;
+  if (aux_labels != nullptr) {
+    K2_CHECK(IsCompatible(fsa, *aux_labels));
+    K2_CHECK_EQ(aux_labels->Dim(), arcs.Dim());
+    p = aux_labels->Data();
+  }
+  float scale = 1;
+  if (negate_scores) scale = -1;
+
+  std::ostringstream os;
+
+  int32_t n = arcs.Dim();
+  char sep = ' ';
```

Comment:
`kDelim`?

Reply:
For printing, I described in the documentation that fields are separated by only ONE space.
For parsing, I also described that it supports multiple spaces/tabs between fields.